plc_ioc_status.py

```diff
@@ -0,0 +1,182 @@
+import functools
+import itertools
+from string import Template
+
+from pydm import Display
+from pydm.widgets import PyDMLabel
+from pydm.widgets.byte import PyDMBitIndicator
+from pydm.widgets.channel import PyDMChannel
+from qtpy import QtWidgets
+from qtpy.QtGui import QColor
+
+
+class PLCIOCStatus(Display):
+    _on_color = QColor(0, 255, 0)
+    _off_color = QColor(100, 100, 100)
+
+    def __init__(self, parent=None, args=None, macros=None):
+        super(PLCIOCStatus, self).__init__(
+            parent=parent, args=args, macros=macros)
+        self.config = macros
+        self.ffs_count_map = {}
+        self.ffs_label_map = {}
+        self.setup_ui()
+
+    def setup_ui(self):
+        self.setup_plc_ioc_status()
+
+    def setup_plc_ioc_status(self):
+        ffs = self.config.get('fastfaults')
+        if not ffs:
+            return
+        if self.plc_ioc_container is None:
+            return
+
+        for ff in ffs:
+            prefix = ff.get('prefix')
+            ffo_start = ff.get('ffo_start')
+            ffo_end = ff.get('ffo_end')
+            ff_start = ff.get('ff_start')
+            ff_end = ff.get('ff_end')
+
+            ffos_zfill = len(str(ffo_end)) + 1
+            ffs_zfill = len(str(ff_end)) + 1
+            entries = itertools.product(
+                range(ffo_start, ffo_end + 1),
+                range(ff_start, ff_end + 1)
+            )
+
+            plc_name = prefix.strip(':')
+            plc_macros = dict(P=prefix)
+            # get the heartbeat of the PLC to
+            heart_ch = Template(
+                'ca://${P}HEARTBEAT').safe_substitute(**plc_macros)
+
+            label_name = QtWidgets.QLabel(str(plc_name))
+            label_online = QtWidgets.QLabel()
+            label_in_use = QtWidgets.QLabel()
+            label_alarmed = QtWidgets.QLabel()
+            label_heartbeat = PyDMLabel(init_channel=heart_ch)
+
+            # if alarm of label_hearbeat == INVALID == plc down
+            # if cannot connect to label_heartbeat = ioc down
+            status_channel = PyDMChannel(
```

Comment:
Not sure if I needed to add this `PyDMChannel` here, I think I should be able to catch the signal emitted when severity changes directly from the `label_heartbeat`?

Reply:
I think your intuition is correct, but I'm having trouble figuring out exactly how to do it.
I think as implemented there is an issue where you're actually checking the severity of the severity field... I'm not sure if that is a reasonable value or not. This should either be the value of the severity field or the severity of the base PV.

Reply:
But I think this implementation is good as-is

Reply:
It just occurred to me that this particular PV's severity isn't actually going to tell us anything about the PLC status, since it isn't querying the PLC for any data. I'll need to find a different PV to use here.

Reply:
Here's a fun suggestion:
```
PLC:LFE:MOTION:TaskInfo:1:CycleCount
```
with `PREFIX=PLC:LFE:MOTION:` here

This is a number that increments every time the PLC finishes a processing cycle, and is queried from the PLC itself. So:
- If this number is increasing, the PLC is running and is online
- If this number is static, the PLC stopped (but is still online)
- If the severity goes to INVALID, the plc is offline

Some facts about the EPICS PV:
- It refreshes about once per second
- It is 32 bit unsigned, so it maxes out at 4294967295. For the fastest PLCs we have that will take 49.71 days to overflow, which could feasibly happen, so resetting to zero at the top and counting up again is expected behavior.

It would be good to show this number in a new column as a diagnostic, and you could also use it in the summary circle (though technically just showing the number is sufficient)

Reply:
Cool thanks so much for looking into this, I will use this PV instead. Also thanks for the EPICS PV info.

Reply:
I think it would be nice to have like a third state of the PLC Status - that would indicate when PLC is Online & Stopped - for this I would have to have like a timeout that will trigger after some time (if the cycle count does not get updated) - i don't know if we need this though, since we can observe if the counter is updating and if the little label has any alarms.

This is kind of how it looks when PLC Cycle Count.SEVR == INVALID:
![plc_status_off](https://user-images.githubusercontent.com/62306310/111700446-a717ee00-87f6-11eb-9684-0af91018a404.PNG)

and when PLC Cycle Count.SEVR != INVALID:
![plc_status_on](https://user-images.githubusercontent.com/62306310/111700459-aaab7500-87f6-11eb-80ad-5b8d76ccabc0.PNG)

Reply:
I agree about the third state, but I didn't request it because it seemed like unnecessary complexity for a nice-to-have, since the information is a bit redundant. If you think it's easy/quick to implement then feel free to go for it, otherwise I think it's better to spend your efforts on other things.

Reply:
Also, I like how those screenshots look

Reply:
I don't know how to implement it yet - I am assuming I could use a `QTimer ` but I don't know how this works with python, i have used it a while ago with C++. I think we can add it later if we want to - because it might take some time for me.

Reply:
I need to fix the Cycle PV here - I think it is not working on pcds-dev